kernel/solo5.h

```diff
@@ -136,8 +165,13 @@ uint64_t solo5_clock_monotonic(void);
 uint64_t solo5_clock_wall(void);
 
 /*
- * Blocks until monotonic time reaches until_nsecs or I/O is possible,
- * whichever is sooner. Returns 1 if I/O is possible, otherwise 0.
+ * Blocks until monotonic time reaches until_nsecs or network I/O is possible,
+ * whichever is sooner.
+ *
+ * If network I/O is possible:
+ *     Returns 1.
+ * If monotonic time has reached until_nsecs and network I/O is not possible:
+ *     Returns 0.
  *
  * TODO: Extend this interface to select which I/O events are of interest.
  */
```

Comment:
In my experience, the use of relative timeouts in POSIX poll() leads to unnecessary complications in the common case for an event loop, which is when I/O events are regularly arriving _and_ the application has some timers scheduled. Ignoring the Mirage implementation for now, consider the two options in this scenario:

_If relative timeouts are used_: The application _must_ recalculate the relative _timeout_ passed to poll() in each iteration of the loop. Given that poll() does not return[1] the remaining time, this involves a call to get the current time.

_If absolute timeouts are used_: The application can keep calling poll(_t_) in each iteration of the loop, where _t_ is the nearest scheduled timer, until poll() indicates that the time _t_ has been reached[2], at which point the application selects the next scheduled timer.

[1] If poll() is implemented in the backend using select() then some implementations _do_ return the time not slept. However, POSIX permits either behaviour so in practice this cannot be relied upon.

[2] Our API does not currently clearly indicate if the timeout has expired. This is a bug. Here is a strawman proposal for a fixed interface (without any ability to select what I/O is interesting since we don't have any infrastructure for that yet):

````
/*
 * Wait until network packets are available or monotonic time reaches (deadline), whichever is sooner.
 *
 * On success:
 *    Returns 0, and (*io) is true if network packets are available, (*timed_out) is true if (deadline) has passed.
 * On failure:
 *    Returns -1.
 */
int solo5_wait(uint64_t deadline, bool *io, bool *timed_out);
````

In the above proposal I've deliberately renamed the call to `solo5_wait()` to disassociate it from the POSIX poll(); I've had several conversations with people who were confused by the "poll" in the name.

Note that the change in semantics (reporting whether or not timeout has expired) requires a bit more work from the (ukvm) backends. That's fine and can be optimized later using epoll()/timerfd_create() on Linux and kqueue() on FreeBSD.

In summary, I'd prefer that we stick with absolute timeouts for poll() and implement the above proposal or similar as part of this change.

Aside: I went hunting for some modern "wait for events" APIs while thinking about this, those [implemented](https://github.com/fuchsia-mirror/magenta/blob/master/docs/syscalls/waitset_wait.md) by Magenta (the kernel used in Google Fuchsia) are a useful source of future inspiration and incidentally exclusively use absolute timeouts.